_posts/2019-07-23-extracting-fsspec-from-dask.md

```diff
@@ -0,0 +1,133 @@
+---
+layout: post
+title: Extracting fsspec from Dask
+author: Martin Durant
+tags: [IO]
+draft: true
+theme: twitter
+---
+{% include JB/setup %}
+
+
+Summary
+-------
+
+Over the past few years, Dask's IO capability has grown gradually and organically, to
+include a number of file-formats, and the ability to access data seamlessly on various
+remote/cloud data systems. This has been achieved through a number of sister packages
+for viewing cloud resources as file-systems, and dedicated code in the `bytes`
+subpackage within Dask. For the sake of consolidating the behaviours of the
+various backends, providing a single reference specification for any new backends,
+and to make this set of file-system operations available even without Dask, I
+created `fsspec`. This last week, Dask changed to use `fsspec` directly for its
+IO needs, and I would like to describe in detail here the benefits of this change.
+
+History
+-------
+
+The first file-system I wrote was [`hdfs3`](https://github.com/dask/hdfs3), a think wrapper
+around the `libhdfs3` C library. At the time, Dask had acquired the ability to run on a
+distributed cluster, and HDFS was the most popular storage solution for these (in the
+commercial world, at least), so a solution was required. The python API closely matched
+the C one, which is turn followed the Java API and posix standards.
+
+Pretty soon, it became apparent that cloud resources would be at least as important as in-cluster
+file-systems, and so followed [s3fs](https://github.com/dask/s3fs), 
+[adlfs](https://github.com/Azure/azure-data-lake-store-python), and [gcsfs](https://github.com/dask/gcsfs).
+Each followed the same pattern, but with some specific code for the given interface, and
+improvements based on the experience of the previous interfaces. During this time, Dask's
+needs also evolved, due to more complex file formats such as parquet. Code to interface to
+the different backends and adapt their methods ended up in the Dask repository.
+
+In the meantime, other file-system interfaces arrived, particularly 
+[`pyarrow`'s](https://arrow.apache.org/docs/python/filesystems.html), which had its own HDFS
+implementation and direct parquet reading. But we would like all of the tools in
+the ecosystem to work together well, so that Dask can read parquet using either
+engine from any of the storage backends.
+
+Code duplication
+----------------
+
+Copying an interface, adapting it and releasing it, as I did with each iteration of the file-system,
+is certainly a quick way to get a job done. However, when you then want to change the behaviour, or
+add new functionality, it turns out you need to repeat the work in each place (not very DRY) or have
+the interfaces diverge slowly. Good examples of this were `glob` and `walk`, which supported various
+options for the former, and returned different things (list, versions dir/files iterator) for the
+latter.
```

Comment:
This might be a good point to include a small code example.  I'm not sure how important the code actually is, but at this point you've only got prose here, and the reader's eyes might appreciate a different style brieflly.